p-refinement

wthrowe opened the pull request:
This does not include methods of actually making refined domains, but includes everything you would need to evolve on such a domain if you had one.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add 0-dim case to apply_matrices

Variables are projected on boundaries, which can be 0-dimensional.

wthrowe committed:
Add mortar meshes to the DataBox

wthrowe committed:
Add support for p-refinement to ABFGTS action

wthrowe committed:
Add functions for moving data to and from mortars

wthrowe committed:
Split LocalData into a struct

It turns out the different quantities are on different meshes, so they
can't all be stored together in a Variables.

wthrowe committed:
Add p-refinement to SendDataForFluxes

wthrowe committed:
Instantiate CoordinateMaps::Identity<3>

It is useful for tests where you don't care what the map is.

nilsvu commented on src/NumericalAlgorithms/DiscontinuousGalerkin/FluxCommunicationTypes.hpp:
```diff
@@ -39,18 +44,24 @@ struct FluxCommunicationTypes {
+    /// PackagedData on the mortar mesh
+    PackagedData packaged_data;
+    /// Unit normal dotted into the fluxes on the face mesh
+    db::item_type<normal_dot_fluxes_tag> normal_dot_fluxes;
```
The `PackageTags` can/will include the `normal_dot_fluxes` or a subset of them, so perhaps we should keep them in one Variables to not duplicate them

wthrowe replied:
At the moment they are on different meshes, so that wouldn't work.  However, it might be better to change the code to do the calculations with the local n.F on the mortar.  It would make the sending code and the flux-lifting code more complicated, but it would remove this extra storage and simplify the upcoming h-refinement logic.  I will try it and see which seems better.

nilsvu replied:
Aren't we storing both on every mortar to do the flux-lifting? Could you explain to me why they are on different meshes?

wthrowe replied:
The current calculation being done is roughly `n (P F* - F)` with `F*` on the mortar and `n` and `F` on the face, with `P` the projection operator.  The second method I'm considering is `n P (F* - F)`, which moves `F` to the mortar.  The two expressions are equivalent in the absence of h-refinement.  (The multiplication by the magnitude of the normal must be done on the face mesh.)  The quantities not needed to calculate `F*` are only stored locally, not sent to the neighboring element.

It is easier to collect the data together for the first method, because the mortar-mesh data on both sides is the same (whatever is needed to compute `F*`), and in some cases we can avoid projecting the fluxes to the mortar.  The second method avoids double storage of fluxes in the common case where they are needed for `F*`, and I believe leads to a simpler expression for h-refinement.

wthrowe committed:
Move flux subtraction out of dg::lift_flux

We will want to do a projection between the subtraction and lifting to
the volume.

wthrowe commented:
I think I'm going to need some help with the OSX issue.  The test in 
ccc14c17cadea91c80ab427e8ba0e407b74208b9 should demonstrate the problem sufficiently.

kidder commented:
okay, I will see if I can reproduce on my laptop...

kidder commented:
@wthrowe @nilsdeppe the good news is I can reproduce the error on my mac...
the bad news is that 
`std::array<NonDefaultConstructible,0> bla{{}};` 
produces...
`/Library/Developer/CommandLineTools/usr/include/c++/v1/array:137:16: note: 
      default constructor of 'array<NonDefaultConstructible, 0>' is implicitly
      deleted because field '__elems_' has a deleted default constructor
    value_type __elems_[_Size > 0 ? _Size : 1];`
as it appears to be making a c-style array of size 1 internally 😱 
In fact, `std::array<NonDefaultConstructible,0> bla{{ndc}};` compiles (but of course fails on other compilers...)

wthrowe commented:
[Found it.](https://bugs.llvm.org/show_bug.cgi?id=35491)  I don't know whether that's been picked up in any releases yet, and I don't know how closely Apple follows upstream.

fmahebert commented:
In the past, I remember it was possible to dig around in certain documentation files and find the LLVM version on which Apple's clang was based. Now, I no longer know how to find this information. Maybe @nilsleiffischer knows more about this.

wthrowe commented:
After the above discussion with @nilsleiffischer, I have modified this PR to store the local flux on the mortar mesh.  Both methods seem similarly complex, but I chose this version because it makes the h-refinement logic (not in this PR) somewhat more straightforward.

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyBoundaryFluxesGlobalTimeStepping.hpp:
```diff
@@ -100,8 +100,8 @@ struct ApplyBoundaryFluxesGlobalTimeStepping {
 
-            const auto data = this_mortar_data.second.extract();
-            const auto& local_mortar_data = data.first;
+            auto data = this_mortar_data.second.extract();
```
check this change

nilsdeppe replied:
Was this checked?

wthrowe replied:
Yes.  It's been squashed into the appropriate commit.

kidder reviewed (commented):
@nilsdeppe please look at

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp:
```diff
@@ -0,0 +1,72 @@
+                                    const Mesh<Dim>& face_mesh,
+                                    const Mesh<Dim>& mortar_mesh) noexcept {
+  const Matrix identity{};
+  auto projections = make_array<Dim>(std::cref(identity));
```
Can you name this `projection_matrices` or something descriptive?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp:
```diff
@@ -0,0 +1,72 @@
+}
+
+/// \ingroup DiscontinuousGalerkinGroup
+/// Project variables from a mortar to a face.
```
Can you document what the `face_mesh` vs the `mortar_mesh` is?

wthrowe replied:
I don't really have more to say than what's in the variable name: `face_mesh` is the face's mesh, `mortar_mesh` is the mortar's mesh.  What do you feel needs clarification?

nilsdeppe replied:
I was thinking of of explaining what face and mortar are, but this isn't the right place. We'll deal with where to discuss and document that later :)

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp:
```diff
@@ -0,0 +1,72 @@
+namespace dg {
+/// \ingroup DiscontinuousGalerkinGroup
+/// Compute the appropriate mesh for use on a mortar given the meshes
+/// of the faces on the two sides.
```
Can you define what `appropriate` means?

wthrowe commented:
Rebased.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -0,0 +1,136 @@
+  CHECK(dg::mortar_mesh(lgl_mesh<1>({{3}}), lgl_mesh<1>({{5}})) ==
+        lgl_mesh<1>({{5}}));
+  CHECK(dg::mortar_mesh(lgl_mesh<2>({{2, 5}}), lgl_mesh<2>({{3, 4}})) ==
+        lgl_mesh<2>({{3, 5}}));
```
Could you check a 3D case as well?

wthrowe replied:
I don't even know why I instantiated this for 3D.  We shouldn't need 3D mortars.  I've removed the instantiation instead.

nilsdeppe replied:
Sounds good!

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -0,0 +1,136 @@
+
+SPECTRE_TEST_CASE("Unit.DG.MortarHelpers.projections",
+                  "[Unit][NumericalAlgorithms]") {
+  {
```
Could you add a 1-line comment at the beginning of the scopes summarizing what they're checking? E.g. `// Check single grid point`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -0,0 +1,136 @@
+      vars.initialize(mortar_mesh.number_of_grid_points());
+      get(get<Var>(vars)) = func(mortar_coords);
+      if (face_mesh.extents(0) < mortar_mesh.extents(0)) {
+        // This is the face_mesh[0]th basis function (5 here).  There
```
Could you add to this comment where this term arises? Just something short, e.g. `// Add term from X`, mostly to act as a refresher when reading or doing source transformations. (Same below where this happens again.)

I like that you point to the bug! :)

nilsdeppe commented on src/Evolution/Executables/ScalarWave/CMakeLists.txt:
```diff
@@ -43,6 +43,7 @@ function(add_scalar_wave_executable DIM)
   target_link_libraries(
```
Could you elaborate on "working with" in the commit message?

nilsdeppe reviewed (changes requested):
I've only had time to look at the first two commits but wanted to get the feedback in sooner rather than later.

Note to me:
- Add 0-dim case to apply_matrices
- Add helper functions for working with mortars

wthrowe commented:
@nilsdeppe could you please complete your review?  We are on a fairly tight deadline.

nilsdeppe reviewed (dismissed):
I'm happy for you to squash :)

nilsdeppe commented:
Only the dox test failed, moving on...
